encodings/internal.js

```diff
@@ -196,3 +198,34 @@ InternalDecoderCesu8.prototype.end = function() {
         res += this.defaultCharUnicode;
     return res;
 }
+
+//------------------------------------------------------------------------------
+// check the chunk boundaries for surrogate pair
+
+function InternalEncoderUtf8(options, codec) {
+    this.lowSurrogate = '';
+}
+
+InternalEncoderUtf8.prototype.write = function (str) {
+    var lastr = str[str.length - 1]
+    if ('\uD800' < lastr && lastr <= '\uDBFF') {
+        if (this.lowSurrogate) {
+            str = this.lowSurrogate + str
+        }
+        this.lowSurrogate = lastr
+        str = str.slice(0, str.length - 1)
+        return Buffer.from(str, this.enc);
+    } else {
+        var buf = Buffer.from(this.lowSurrogate + str, this.enc);
+        this.lowSurrogate = ''
+        return buf
+    }
```

Comment:
I *think* this should work correctly, but can we streamline it even more?
 1. Can you make this function have only one `Buffer.from(` and one string concatenation `this.lowSurrogate + str`? The branches of this if statement have some repetition in them and I think removing it would make the code clearer.
 2. `'\uD800' < lastr && lastr <= '\uDBFF'` is a string comparison and is afaik expensive. Can we maybe make `lastr` and `this.lowSurrogate` integers using `s.charCodeAt` and switch to an integer comparison? Also could you name `lastr` better? I'm not sure what the current name means.
 3. Please make sure the function does not throw if str is empty.